Duplicati/Library/Backend/OneDrive/MicrosoftGraphBackend.cs

```diff
@@ -796,6 +943,119 @@ private T ParseResponse<T>(HttpWebResponse response)
+            }
+
+            throw new UploadSessionException(createSessionResponse, fragment, fragmentCount, ex);
+        }
+
+        private void SetRetryAfter(RetryConditionHeaderValue retryAfter)
+        {
+            if (retryAfter != null)
+            {
+                DateTimeOffset? delayUntil = null;
+                if (retryAfter.Delta.HasValue)
+                {
+                    delayUntil = DateTimeOffset.UtcNow + retryAfter.Delta.Value;
+                }
+                else if (retryAfter.Date.HasValue)
+                {
+                    delayUntil = retryAfter.Date.Value;
+                }
+
+                if (delayUntil.HasValue)
+                {
+                    // Set the retry timestamp to the UTC version of the timestamp.
+                    long newRetryAfter = delayUntil.Value.UtcTicks;
+
+                    // Update the persisted retry after timestamp
+                    long replacedRetryAfter;
+                    long currentRetryAfter;
+                    do
+                    {
+                        currentRetryAfter = Interlocked.Read(ref this.retryAfter);
+
+                        if (newRetryAfter < currentRetryAfter)
+                        {
+                            // If the current retry after is already past the new value, then no need to update it again.
+                            break;
+                        }
+                        else
+                        {
+                            replacedRetryAfter = Interlocked.CompareExchange(ref this.retryAfter, newRetryAfter, currentRetryAfter);
+                        }
+                    }
+                    while (replacedRetryAfter != currentRetryAfter);
+                }
+            }
+        }
+
+        private void WaitForRetryAfter()
+        {
+            this.WaitForRetryAfter(CancellationToken.None).Await();
+        }
+
+        private async Task WaitForRetryAfter(CancellationToken cancelToken)
+        {
+            // Make sure this is thread safe in case multiple calls are made concurrently to this backend
+            // This is done by reading the value into a local value which is then parsed and operated on locally.
+            long retryAfterTicks = Interlocked.Read(ref this.retryAfter);
+            DateTimeOffset delayUntil = new DateTimeOffset(retryAfterTicks, TimeSpan.Zero);
+
+            TimeSpan delay;
+            DateTimeOffset now = DateTimeOffset.UtcNow;
+
+            // Make sure delayUntil is in the future and delay until then if so
+            if (delayUntil >= now)
+            {
+                delay = delayUntil - now;
+            }
+            else
+            {
+                // If the date given was in the past then don't wait at all
+                delay = TimeSpan.Zero;
+            }
+
+            if (delay > TimeSpan.Zero)
+            {
+                await Task.Delay(delay).ConfigureAwait(false);

```

Comment:
Would it help to have a logging statement here to help debug jobs that appear to be "stalled"?

Reply:
Adding logging to these backends isn't a bad idea - that's something I didn't really look into when originally adding them. It might be worth adding some messages to more than just here though. I'll take a look (probably next week).

Reply:
Some logging has been added.